Add maxDecompress, maxFileWrites, separate download, decompres and file write operation to improve network utilization

joshuatam opened the pull request:
Add maxFileWrites, separate download and file write operation to improve network utilization.

### Description
When using GN 0.6.0 to download the game, the download time is not very good as each of the download job is bounded to the file write operation. Sepearating file write operation and download operation could make the download process smoother and faster.

### Checklist
- [x] Code compiles correctly
- [x] All tests passing
- [ ] Samples run successfully
- [x] Extended the README / documentation, if necessary

joshuatam committed:
Add maxFileWrites, separate download and file write operation to improve network utilization

utkarshdalal commented:
It would be good to see how much heat this causes. If it's not heating up the device, this is a crazy improvement!

joshuatam committed:
change notifyListeners to use coroutineScope for non blocking call to update UI

joshuatam commented:
> It would be good to see how much heat this causes. If it's not heating up the device, this is a crazy improvement!

Pretty much the same temperature I observed, still need your verification lol.

And as my device only got 8 cores, using lower value than the cpu core number could always give rooms for more download running.

utkarshdalal commented:
I'm not able to approve PRs in this repo so we'll have to wait for Lossy

Utkarsh

On Tue, 9 Dec, 2025, 21:12 Joshua Tam, ***@***.***> wrote:

> *joshuatam* left a comment (LossyDragon/JavaSteam#6)
> <https://github.com/LossyDragon/JavaSteam/pull/6#issuecomment-3632928546>
>
> It would be good to see how much heat this causes. If it's not heating up
> the device, this is a crazy improvement!
>
> Pretty much the same temperature I observed, still need your verification
> lol.
>
> And as my device only got 8 cores, using lower value than the cpu core
> number could always give rooms for more download running.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/LossyDragon/JavaSteam/pull/6#issuecomment-3632928546>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4SWC2TP4O6BDAR2XYEYF34A3UWNAVCNFSM6AAAAACOOFZAJSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMMZSHEZDQNJUGY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

joshuatam committed:
changed to use pre-launch file workers for better I/O performance

LossyDragon commented on javasteam-depotdownloader/src/main/kotlin/in/dragonbra/javasteam/depotdownloader/DepotDownloader.kt:
```diff
@@ -1564,8 +1528,10 @@ class DepotDownloader @JvmOverloads constructor(
-    private fun notifyListeners(action: (IDownloadListener) -> Unit) {
-        listeners.forEach { listener -> action(listener) }
+    private suspend fun notifyListeners(action: (IDownloadListener) -> Unit) {
+        coroutineScope {
```
Would `scope.launch{}` work here too?

`scope.launch(Dispatchers.Main){}` if it needs to notify on the main thread.

joshuatam replied:
Thanks, updated

joshuatam replied:
Not working, tested, running on main thread cannot complete the whole download process.

joshuatam committed:
update notifyListeners to run on main thread instead launching other new thread

joshuatam commented:
@LossyDragon Actually another idea just pop up, I think maybe we can just spawn N threads and use the receiveAsFlow to let it runs automatically.

So every thread will process: 1 download task + 1 save task.

In ideal situation, internet throughput can be optimized as no need to wait other download thread complete.

joshuatam committed:
Revert "update notifyListeners to run on main thread instead launching other new thread"

This reverts commit 110abe939ed2c7afba22d9636c99295815358d39.

joshuatam committed:
Improves concurrent depot downloading

Refactors depot downloading to use separate coroutines for network chunk processing and file writing, improving concurrency and overall download speed.

This change introduces dedicated coroutine workers for network chunk downloads and file writing, managing these tasks in parallel. It also ensures proper cleanup and channel closing, preventing resource leaks and ensuring all workers complete their tasks before the download is considered finished. Additionally, it improves the handling of last depot scenarios.

LossyDragon committed:
Lint formatting

joshuatam committed:
Refactors chunk processing with Flows

Migrates from Channels to Flows for chunk processing to
improve concurrency and simplify data flow management.

This change introduces dedicated flows for decompression and
file writing, enabling better control over resource utilization
and error handling. It also decouples network I/O from
decompression and writing operations by processing in parallel,
preventing potential bottlenecks.

Increases chunk processing throughput and reduces latency
by better utilizing available CPU cores.

joshuatam committed:
Updates exception handling in CDN client

Refines exception handling for chunk downloads in the CDN client.

Adjusts the logic to validate destination buffer size based on whether a depot key is provided.

joshuatam committed:
Lint fix: imports specific Kotlin flow

Imports the necessary Kotlin flow operators for the
'DepotDownloader' class to resolve compilation errors
and ensure proper data stream processing.

LossyDragon committed:
Fix lexicographic order

joshuatam committed:
Changes dispatcher for download task

Changes the dispatcher used for the download task to IO.

This ensures that download tasks, which are I/O bound,
run on a dispatcher that is optimized for I/O operations,
preventing potential blocking of the default dispatcher.

joshuatam committed:
Adds parent job cancellation support

Ensures that the downloader is cancelled when its parent job is cancelled, allowing for better control and resource management in complex workflows.
